components/encryption/src/io.rs

```diff
@@ -106,63 +336,50 @@ pub fn create_aes_ctr_crypter(
     Ok((cipher, crypter))
 }
 
-/// Implementation of EncrypterReader and DecrypterReader.
-struct CrypterReader<R> {
-    reader: R,
-
+struct CrypterCore {
     method: EncryptionMethod,
     key: Vec<u8>,
     mode: Mode,
     initial_iv: Iv,
     crypter: Option<OCrypter>,
     block_size: usize,
 
-    crypter_buffer: Vec<u8>,
+    buffer: Vec<u8>,
 }
 
-impl<R> CrypterReader<R> {
-    pub fn new(
-        reader: R,
-        method: EncryptionMethod,
-        key: &[u8],
-        mode: Mode,
-        iv: Option<Iv>,
-    ) -> Result<(CrypterReader<R>, Iv)> {
+impl CrypterCore {
+    pub fn new(method: EncryptionMethod, key: &[u8], mode: Mode, iv: Iv) -> Result<Self> {
         crate::verify_encryption_config(method, key)?;
-        let iv = iv.unwrap_or_else(Iv::new_ctr);
-        Ok((
-            CrypterReader {
-                reader,
-                method,
-                key: key.to_owned(),
-                mode,
-                crypter: None,
-                block_size: 0,
-                initial_iv: iv,
-                crypter_buffer: Vec::new(),
-            },
-            iv,
-        ))
+        Ok(CrypterCore {
+            method,
+            key: key.to_owned(),
+            mode,
+            crypter: None,
+            block_size: 0,
+            initial_iv: iv,
+            buffer: Vec::new(),
+        })
     }
 
-    fn reserve_buffer(&mut self, size: usize) {
+    #[inline]
+    fn reset_buffer(&mut self, size: usize) {
         // OCrypter require the output buffer to have block_size extra bytes, or it will panic.
-        let target = std::cmp::min(size, MAX_CRYPTER_READER_BUFFER_SIZE) + self.block_size;
-        if target > self.crypter_buffer.len() {
-            self.crypter_buffer.resize(target, 0);
+        self.buffer.reserve(size + self.block_size);
+        unsafe {
+            self.buffer.set_len(size + self.block_size);
```

Comment:
not sure if it's okay to not fill the padding of 0. /cc @yiwu-arbug

Reply:
It's just an output buffer, the behavior is stated in document: https://docs.rs/openssl/0.9.1/openssl/symm/struct.Crypter.html#method.update